Sync/Synchronizers/IndependentRecordSynchronizer.swift

```diff
@@ -120,4 +130,36 @@ extension TimestampedSingleCollectionSynchronizer {
                 return deferMaybe(timestamp)
             }
     }
+
+    func uploadRecordsSingleBatch<T>(_ records: [Record<T>], lastTimestamp: Timestamp, storageClient: Sync15CollectionClient<T>) -> DeferredTimestamp {
+        if records.isEmpty {
+            log.debug("No modified records to upload.")
+            return deferMaybe(lastTimestamp)
+        }
+
+        if records.count > storageClient.maxBatchPostRecords {
+            return deferMaybe(TooManyRecordsError())
+        }
+
+        var additionalPosts = 0
+        func onCollectionUploaded(result: POSTResult, timestamp: Timestamp?) -> DeferredTimestamp {
+            additionalPosts += 1
+            return deferMaybe(timestamp ?? 0)
+        }
+
+        let batch = storageClient.newBatch(ifUnmodifiedSince: (lastTimestamp == 0) ? nil : lastTimestamp, onCollectionUploaded: onCollectionUploaded)
+        return batch.addRecords(records)
+            >>== { _ in
+                return batch.flush()
+            } >>== { resp -> Success in
+                guard resp.value.success.count == records.count && additionalPosts == 0 else {
```

Comment:
This is a bit too much like digging around in the guts of the uploader for my tastes. You want two things:

- To know whether all records were uploaded.
- To only commit (and, ideally, to only upload) if all of the records can fit in a single _commit_.

It's possible for even the batch API to span commits (more than 10K records, IIRC), which would mean sadness for us.

I think this logic could productively belong in the uploader, no?